README.md
![lingua](images/logo.png) 

<br>

[![ci build status][ci badge]][ci url]
[![codecov][codecov badge]][codecov url]
[![supported languages][supported languages badge]](#supported-languages)
[![Kotlin platforms badge][Kotlin platforms badge]][Kotlin platforms url]
[![license badge][license badge]][license url]

[![javadoc][javadoc badge]][javadoc url]
[![Maven Central][Maven Central badge]][Maven Central]
[![Download][lingua version badge]][lingua download url]

---
### Quick Info
* this library tries to solve language detection of very short words and phrases, even shorter than tweets
* makes use of both statistical and rule-based approaches
* outperforms *Apache Tika*, *Apache OpenNLP* and *Optimaize Language Detector* for more than 70 languages
* works within every Java 6+ application
* no additional training of language models necessary
* api for adding your own language models
* offline usage without having to connect to an external service or API
* can be used in a REPL for a quick try-out
---

## 1. What does this library do?
Its task is simple: It tells you which language some provided textual data is written in. 
This is very useful as a preprocessing step for linguistic data in natural language 
processing applications such as text classification and spell checking. 
Other use cases, for instance, might include routing e-mails to the right geographically 
located customer service department, based on the e-mails' languages.

## 2. Why does this library exist?
Language detection is often done as part of large machine learning frameworks or natural 
language processing applications. In cases where you don't need the full-fledged 
functionality of those systems or don't want to learn the ropes of those, 
a small flexible library comes in handy. 

So far, three other comprehensive open source libraries working on the JVM for this task 
are [Apache Tika], [Apache OpenNLP] and [Optimaize Language Detector]. 
Unfortunately, especially the latter has three major drawbacks:
 
1. Detection only works with quite lengthy text fragments. 
For very short text snippets such as Twitter messages, it doesn't provide adequate results.
2. The more languages take part in the decision process, the less accurate are the detection results.
3. Configuration of the library is quite cumbersome and requires some knowledge about the statistical 
methods that are used internally.

*Lingua* aims at eliminating these problems. It nearly doesn't need any configuration and 
yields pretty accurate results on both long and short text, even on single words and phrases. 
It draws on both rule-based and statistical methods but does not use any dictionaries of words. 
It does not need a connection to any external API or service either. 
Once the library has been downloaded, it can be used completely offline. 

## 3. Which languages are supported?

Compared to other language detection libraries, *Lingua's* focus is on *quality over quantity*, that is, 
getting detection right for a small set of languages first before adding new ones. 
Currently, the following 75 languages are supported:

- A
  - Afrikaans
  - Albanian
  - Arabic
  - Armenian
  - Azerbaijani
- B
  - Basque
  - Belarusian
  - Bengali
  - Norwegian Bokmal
  - Bosnian
  - Bulgarian
- C
  - Catalan
  - Chinese
  - Croatian
  - Czech
- D
  - Danish
  - Dutch
- E
  - English
  - Esperanto
  - Estonian
- F
  - Finnish
  - French
- G
  - Ganda
  - Georgian
  - German
  - Greek
  - Gujarati
- H
  - Hebrew
  - Hindi
  - Hungarian
- I
  - Icelandic
  - Indonesian
  - Irish
  - Italian
- J
  - Japanese
- K
  - Kazakh
  - Korean
- L
  - Latin
  - Latvian
  - Lithuanian
- M
  - Macedonian
  - Malay
  - Maori
  - Marathi
  - Mongolian
- N
  - Norwegian Nynorsk
- P
  - Persian
  - Polish
  - Portuguese
  - Punjabi
- R
  - Romanian
  - Russian
- S
  - Serbian
  - Shona
  - Slovak
  - Slovene
  - Somali
  - Sotho
  - Spanish
  - Swahili
  - Swedish
- T
  - Tagalog
  - Tamil
  - Telugu
  - Thai
  - Tsonga
  - Tswana
  - Turkish
- U
  - Ukrainian
  - Urdu
- V
  - Vietnamese
- W
  - Welsh
- X
  - Xhosa
- Y
  - Yoruba
- Z
  - Zulu   

## 4. How good is it?

*Lingua* is able to report accuracy statistics for some bundled test data available for each supported language. 
The test data for each language is split into three parts:
1. a list of single words with a minimum length of 5 characters
2. a list of word pairs with a minimum length of 10 characters
3. a list of complete grammatical sentences of various lengths

Both the language models and the test data have been created from separate documents of the [Wortschatz corpora] 
offered by Leipzig University, Germany. Data crawled from various news websites have been used for training, 
each corpus comprising one million sentences. For testing, corpora made of arbitrarily chosen websites have been used, 
each comprising ten thousand sentences. From each test corpus, a random unsorted subset of 1000 single words, 
1000 word pairs and 1000 sentences has been extracted, respectively.

Given the generated test data, I have compared the detection results of *Lingua*, *Apache Tika*, *Apache OpenNLP* and 
*Optimaize Language Detector* using parameterized JUnit tests running over the data of *Lingua's* supported 75 languages. 
Languages that are not supported by the other libraries are simply ignored for those during the detection process.

Each of the following sections contains two plots. The bar plot shows the detailed accuracy results for each supported 
language. The box plot illustrates the distributions of the accuracy values for each classifier. The boxes themselves 
represent the areas which the middle 50 % of data lie within. Within the colored boxes, the horizontal lines mark the 
median of the distributions.

### 4.1 Single word detection

<br/>

<img src="https://raw.githubusercontent.com/pemistahl/lingua/main/images/plots/boxplot-single-words.png" alt="Single Word Detection Performance" />

<br/>

<details>
    <summary>Bar plot</summary>
    <img src="https://raw.githubusercontent.com/pemistahl/lingua/main/images/plots/barplot-single-words.png" alt="Single Word Detection Performance" />
</details>

<br/><br/>

### 4.2 Word pair detection

<br/>

<img src="https://raw.githubusercontent.com/pemistahl/lingua/main/images/plots/boxplot-word-pairs.png" alt="Word Pair Detection Performance" />

<br/>

<details>
    <summary>Bar plot</summary>
    <img src="https://raw.githubusercontent.com/pemistahl/lingua/main/images/plots/barplot-word-pairs.png" alt="Word Pair Detection Performance" />
</details>

<br/><br/>

### 4.3 Sentence detection

<br/>

<img src="https://raw.githubusercontent.com/pemistahl/lingua/main/images/plots/boxplot-sentences.png" alt="Sentence Detection Performance" />

<br/>

<details>
    <summary>Bar plot</summary>
    <img src="https://raw.githubusercontent.com/pemistahl/lingua/main/images/plots/barplot-sentences.png" alt="Sentence Detection Performance" />
</details>

<br/><br/>

### 4.4 Average detection

<br/>

<img src="https://raw.githubusercontent.com/pemistahl/lingua/main/images/plots/boxplot-average.png" alt="Average Detection Performance" />

<br/>

<details>
    <summary>Bar plot</summary>
    <img src="https://raw.githubusercontent.com/pemistahl/lingua/main/images/plots/barplot-average.png" alt="Average Detection Performance" />
</details>

<br/><br/>

### 4.5 Mean, median and standard deviation

The table below shows detailed statistics for each language and classifier
including mean, median and standard deviation.

<details>
    <summary>Open table</summary>
    <table>
    <tr>
        <th>Language</th>
        <th colspan="5">Average</th>
        <th colspan="5">Single Words</th>
        <th colspan="5">Word Pairs</th>
        <th colspan="5">Sentences</th>
    </tr>
    <tr>
        <th></th>
        <th>Lingua<br>(high accuracy mode)</th>
        <th>Lingua<br>(low accuracy mode)</th>
        <th>&nbsp;&nbsp;Tika&nbsp;&nbsp;</th>
        <th>OpenNLP</th>
        <th>Optimaize</th>
        <th>Lingua<br>(high accuracy mode)</th>
        <th>Lingua<br>(low accuracy mode)</th>
        <th>&nbsp;&nbsp;Tika&nbsp;&nbsp;</th>
        <th>OpenNLP</th>
        <th>Optimaize</th>
        <th>Lingua<br>(high accuracy mode)</th>
        <th>Lingua<br>(low accuracy mode)</th>
        <th>&nbsp;&nbsp;Tika&nbsp;&nbsp;</th>
        <th>OpenNLP</th>
        <th>Optimaize</th>
        <th>Lingua<br>(high accuracy mode)</th>
        <th>Lingua<br>(low accuracy mode)</th>
        <th>&nbsp;&nbsp;Tika&nbsp;&nbsp;</th>
        <th>OpenNLP</th>
        <th>Optimaize</th>
    </tr>
    	<tr>
		<td>Afrikaans</td>
		<td><img src="images/lightgreen.png"> 79</td>
		<td><img src="images/lightgreen.png"> 64</td>
		<td><img src="images/lightgreen.png"> 71</td>
		<td><img src="images/lightgreen.png"> 72</td>
		<td><img src="images/orange.png"> 39</td>
		<td><img src="images/yellow.png"> 58</td>
		<td><img src="images/orange.png"> 38</td>
		<td><img src="images/yellow.png"> 44</td>
		<td><img src="images/yellow.png"> 41</td>
		<td><img src="images/red.png"> 3</td>
		<td><img src="images/green.png"> 81</td>
		<td><img src="images/lightgreen.png"> 62</td>
		<td><img src="images/lightgreen.png"> 70</td>
		<td><img src="images/lightgreen.png"> 75</td>
		<td><img src="images/orange.png"> 22</td>
		<td><img src="images/green.png"> 97</td>
		<td><img src="images/green.png"> 93</td>
		<td><img src="images/green.png"> 98</td>
		<td><img src="images/green.png"> 99</td>
		<td><img src="images/green.png"> 93</td>
	</tr>
	<tr>
		<td>Albanian</td>
		<td><img src="images/green.png"> 88</td>
		<td><img src="images/lightgreen.png"> 80</td>
		<td><img src="images/lightgreen.png"> 79</td>
		<td><img src="images/lightgreen.png"> 71</td>
		<td><img src="images/lightgreen.png"> 70</td>
		<td><img src="images/lightgreen.png"> 69</td>
		<td><img src="images/yellow.png"> 54</td>
		<td><img src="images/yellow.png"> 54</td>
		<td><img src="images/orange.png"> 40</td>
		<td><img src="images/orange.png"> 38</td>
		<td><img src="images/green.png"> 95</td>
		<td><img src="images/green.png"> 86</td>
		<td><img src="images/green.png"> 84</td>
		<td><img src="images/lightgreen.png"> 73</td>
		<td><img src="images/lightgreen.png"> 73</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 99</td>
		<td><img src="images/green.png"> 99</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 98</td>
	</tr>
	<tr>
		<td>Arabic</td>
		<td><img src="images/green.png"> 98</td>
		<td><img src="images/green.png"> 94</td>
		<td><img src="images/green.png"> 97</td>
		<td><img src="images/green.png"> 84</td>
		<td><img src="images/green.png"> 89</td>
		<td><img src="images/green.png"> 96</td>
		<td><img src="images/green.png"> 88</td>
		<td><img src="images/green.png"> 94</td>
		<td><img src="images/lightgreen.png"> 65</td>
		<td><img src="images/lightgreen.png"> 72</td>
		<td><img src="images/green.png"> 99</td>
		<td><img src="images/green.png"> 96</td>
		<td><img src="images/green.png"> 99</td>
		<td><img src="images/green.png"> 88</td>
		<td><img src="images/green.png"> 94</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 99</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 99</td>
		<td><img src="images/green.png"> 100</td>
	</tr>
	<tr>
		<td>Armenian</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/grey.png"> -</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/grey.png"> -</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/grey.png"> -</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/grey.png"> -</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/grey.png"> -</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/grey.png"> -</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/grey.png"> -</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/grey.png"> -</td>
	</tr>
	<tr>
		<td>Azerbaijani</td>
		<td><img src="images/green.png"> 90</td>
		<td><img src="images/green.png"> 82</td>
		<td><img src="images/grey.png"> -</td>
		<td><img src="images/green.png"> 82</td>
		<td><img src="images/grey.png"> -</td>
		<td><img src="images/lightgreen.png"> 77</td>
		<td><img src="images/lightgreen.png"> 71</td>
		<td><img src="images/grey.png"> -</td>
		<td><img src="images/yellow.png"> 60</td>
		<td><img src="images/grey.png"> -</td>
		<td><img src="images/green.png"> 92</td>
		<td><img src="images/lightgreen.png"> 78</td>
		<td><img src="images/grey.png"> -</td>
		<td><img src="images/green.png"> 86</td>
		<td><img src="images/grey.png"> -</td>
		<td><img src="images/green.png"> 99</td>
		<td><img src="images/green.png"> 96</td>
		<td><img src="images/grey.png"> -</td>
		<td><img src="images/green.png"> 99</td>
		<td><img src="images/grey.png"> -</td>
	</tr>
	<tr>
		<td>Basque</td>
		<td><img src="images/green.png"> 84</td>
		<td><img src="images/lightgreen.png"> 74</td>
		<td><img src="images/green.png"> 83</td>
		<td><img src="images/lightgreen.png"> 77</td>
		<td><img src="images/lightgreen.png"> 66</td>
		<td><img src="images/lightgreen.png"> 71</td>
		<td><img src="images/yellow.png"> 56</td>
		<td><img src="images/lightgreen.png"> 64</td>
		<td><img src="images/yellow.png"> 56</td>
		<td><img src="images/orange.png"> 33</td>
		<td><img src="images/green.png"> 87</td>
		<td><img src="images/lightgreen.png"> 76</td>
		<td><img src="images/green.png"> 86</td>
		<td><img src="images/green.png"> 82</td>
		<td><img src="images/lightgreen.png"> 70</td>
		<td><img src="images/green.png"> 93</td>
		<td><img src="images/green.png"> 91</td>
		<td><img src="images/green.png"> 98</td>
		<td><img src="images/green.png"> 92</td>
		<td><img src="images/green.png"> 95</td>
	</tr>
	<tr>
		<td>Belarusian</td>
		<td><img src="images/green.png"> 97</td>
		<td><img src="images/green.png"> 92</td>
		<td><img src="images/green.png"> 96</td>
		<td><img src="images/green.png"> 91</td>
		<td><img src="images/green.png"> 87</td>
		<td><img src="images/green.png"> 92</td>
		<td><img src="images/lightgreen.png"> 80</td>
		<td><img src="images/green.png"> 92</td>
		<td><img src="images/lightgreen.png"> 78</td>
		<td><img src="images/lightgreen.png"> 69</td>
		<td><img src="images/green.png"> 99</td>
		<td><img src="images/green.png"> 95</td>
		<td><img src="images/green.png"> 98</td>
		<td><img src="images/green.png"> 95</td>
		<td><img src="images/green.png"> 92</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 99</td>
	</tr>
	<tr>
		<td>Bengali</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
	</tr>
	<tr>
		<td>Bokmal</td>
		<td><img src="images/yellow.png"> 58</td>
		<td><img src="images/yellow.png"> 49</td>
		<td><img src="images/grey.png"> -</td>
		<td><img src="images/lightgreen.png"> 66</td>
		<td><img src="images/grey.png"> -</td>
		<td><img src="images/orange.png"> 39</td>
		<td><img src="images/orange.png"> 27</td>
		<td><img src="images/grey.png"> -</td>
		<td><img src="images/yellow.png"> 42</td>
		<td><img src="images/grey.png"> -</td>
		<td><img src="images/yellow.png"> 59</td>
		<td><img src="images/yellow.png"> 47</td>
		<td><img src="images/grey.png"> -</td>
		<td><img src="images/lightgreen.png"> 69</td>
		<td><img src="images/grey.png"> -</td>
		<td><img src="images/lightgreen.png"> 75</td>
		<td><img src="images/lightgreen.png"> 74</td>
		<td><img src="images/grey.png"> -</td>
		<td><img src="images/green.png"> 87</td>
		<td><img src="images/grey.png"> -</td>
	</tr>
	<tr>
		<td>Bosnian</td>
		<td><img src="images/orange.png"> 35</td>
		<td><img src="images/orange.png"> 29</td>
		<td><img src="images/grey.png"> -</td>
		<td><img src="images/orange.png"> 26</td>
		<td><img src="images/grey.png"> -</td>
		<td><img src="images/orange.png"> 29</td>
		<td><img src="images/orange.png"> 23</td>
		<td><img src="images/grey.png"> -</td>
		<td><img src="images/red.png"> 12</td>
		<td><img src="images/grey.png"> -</td>
		<td><img src="images/orange.png"> 35</td>
		<td><img src="images/orange.png"> 29</td>
		<td><img src="images/grey.png"> -</td>
		<td><img src="images/orange.png"> 22</td>
		<td><img src="images/grey.png"> -</td>
		<td><img src="images/orange.png"> 40</td>
		<td><img src="images/orange.png"> 36</td>
		<td><img src="images/grey.png"> -</td>
		<td><img src="images/yellow.png"> 44</td>
		<td><img src="images/grey.png"> -</td>
	</tr>
	<tr>
		<td>Bulgarian</td>
		<td><img src="images/green.png"> 87</td>
		<td><img src="images/lightgreen.png"> 78</td>
		<td><img src="images/lightgreen.png"> 73</td>
		<td><img src="images/green.png"> 83</td>
		<td><img src="images/yellow.png"> 48</td>
		<td><img src="images/lightgreen.png"> 70</td>
		<td><img src="images/yellow.png"> 56</td>
		<td><img src="images/yellow.png"> 52</td>
		<td><img src="images/lightgreen.png"> 62</td>
		<td><img src="images/red.png"> 18</td>
		<td><img src="images/green.png"> 91</td>
		<td><img src="images/green.png"> 81</td>
		<td><img src="images/lightgreen.png"> 69</td>
		<td><img src="images/green.png"> 87</td>
		<td><img src="images/orange.png"> 36</td>
		<td><img src="images/green.png"> 99</td>
		<td><img src="images/green.png"> 96</td>
		<td><img src="images/green.png"> 96</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 91</td>
	</tr>
	<tr>
		<td>Catalan</td>
		<td><img src="images/lightgreen.png"> 70</td>
		<td><img src="images/yellow.png"> 58</td>
		<td><img src="images/yellow.png"> 58</td>
		<td><img src="images/yellow.png"> 42</td>
		<td><img src="images/orange.png"> 31</td>
		<td><img src="images/yellow.png"> 51</td>
		<td><img src="images/orange.png"> 33</td>
		<td><img src="images/orange.png"> 32</td>
		<td><img src="images/red.png"> 11</td>
		<td><img src="images/red.png"> 2</td>
		<td><img src="images/lightgreen.png"> 74</td>
		<td><img src="images/yellow.png"> 60</td>
		<td><img src="images/yellow.png"> 57</td>
		<td><img src="images/orange.png"> 32</td>
		<td><img src="images/red.png"> 16</td>
		<td><img src="images/green.png"> 86</td>
		<td><img src="images/green.png"> 81</td>
		<td><img src="images/green.png"> 84</td>
		<td><img src="images/green.png"> 81</td>
		<td><img src="images/lightgreen.png"> 77</td>
	</tr>
	<tr>
		<td>Chinese</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/lightgreen.png"> 69</td>
		<td><img src="images/lightgreen.png"> 78</td>
		<td><img src="images/orange.png"> 31</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/red.png"> 20</td>
		<td><img src="images/orange.png"> 40</td>
		<td><img src="images/red.png"> 0</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 86</td>
		<td><img src="images/green.png"> 94</td>
		<td><img src="images/red.png"> 2</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 100</td>
		<td><img src="images/green.png"> 91</td>
	</tr>
	<tr>
		<td>Croatian</td>
		<td><img src="images/lightgreen.png"> 72</td>
		<td><img src="images/yellow.png"> 60</td>
		<td><img src="images/lightgreen.png"> 74</td>
		<td><img src="images/yellow.png"> 50</td>
		<td><img src="images/yellow.png"> 41</td>
		<td><img src="images/yellow.png"> 53</td>
		<td><img src="images/orange.png"> 36</td>
		<td><img src="images/yellow.png"> 54</td>
		<td><img src="images/orange.png"> 23</td>
		<td><img src="images/red.png"> 8</td>
		<td><img src="images/lightgreen.png"> 74</td>
		<td><img src="images/yellow.png"> 57</td>
		<td><img src="images/lightgreen.png"> 72</td>
		<td><img src="images/yellow.png"> 44</td>
		<td><img src="images/orange.png"> 24</td>
		<td><img src="images/green.png"> 90</td>
		<td><img src="images/green.png"> 85</td>
		<td><img src="images/green.png"> 97</td>
		<td><img src="images/green.png"> 81</td>
		<td><img src="images/green.png"> 91</td>
	</tr>
	<tr>
		<td>Czech</td>
		<td><img src="images/lightgreen.png"> 80</td>
		<td><img src="images/lightgreen.png"> 71</td>
		<td><img src="images/lightgreen.png"> 72</td>
		<td><img src="images/lightgreen.png"> 67</td>

RELEASE_NOTES.md
## Lingua 1.2.2 (released on 02 Aug 2022)

### Bug Fixes

- Due to a bug in the Moshi JSON serialization library, language detection
  was not possible in certain cases. (#144, #147)
- Lingua could not be used properly when a security manager was enabled
  in the JVM. (#141)

## Lingua 1.2.1 (released on 09 Jun 2022)

### Bug Fixes

- An exception was thrown when trying to detect the language of unigrams and
  bigrams in low accuracy mode which operates only with trigrams and larger 
  strings. This has been fixed.

## Lingua 1.2.0 (released on 07 Jun 2022)

### Features

- The library can now be used as a Java 9 module. Thanks to @Marcono1234
  for helping with the implementation. (#120, #138)
- The new method `LanguageDetectorBuilder.withLowAccuracyMode()` has been
  introduced. By activating it, detection accuracy for short text is reduced
  in favor of a smaller memory footprint and faster detection performance. (#136)

### Improvements

- The memory footprint has been reduced significantly by applying several
  internal optimizations. Thanks to @Marcono1234, @fvasco and @sigpwned for
  their help. (#101, #127)
- Several language model files have become obsolete and could be deleted without
  decreasing detection accuracy. This results in a smaller memory footprint and 
  a 36% smaller jar file.

### Bug Fixes

- A bug in the rule engine has been fixed that caused incorrect language
  detection for certain texts. Thanks to @bdecarne who has found it.

### Other changes

- Due to a refactoring of how the internal thread pool works, the method
  `LanguageDetector.destroy()` has been deprecated in favor of the newly
  introduced method `LanguageDetector.unloadLanguageModels()`.

## Lingua 1.1.1 (released on 12 Dec 2021)

### Improvements

- The new method `LanguageDetector.destroy()` has been introduced that frees
  internal resources to prevent memory leaks within application server 
  deployments. (#110, #116)
- Language model loading performance has been improved by creating a manually
  optimized internal thread pool. This replaces the coroutines used in the
  previous release. (#116)

### Bug Fixes

- The character *â* was erroneously not treated as a possible indicator for
  the French language. (#115)
- Language detection was non-deterministic when multiple alphabets had the
  same occurrence count. (105)

## Lingua 1.1.0 (released on 02 May 2021)

### Languages

- There is now support for the Maori language which was contributed to the
  Rust implementation of Lingua. (#93)

### Features

- Language models are now loaded asynchronously and in parallel using Kotlin
  coroutines, making this step more performant. (#84)
- Language Models can now be loaded either lazily (default) or eagerly. (#79)
- Instead of loading multiple copies of the language models into memory for 
  each separate instance of `LanguageDetector`, multiple instances now share
  the same language models and access them asynchronously. (#91)
  
### Improvements

- Language detection for sentences with more than 120 characters now
  performs more quickly by iterating through trigrams only which is
  enough to achieve high detection accuracy.
- Textual input that includes logograms from Chinese, Japanese or Korean
  is now split at each logogram and not only at whitespace. This
  provides for more reliable language detection for sentences that
  include multi-language content. (#85)
  
### Bug Fixes

- For an odd number of words as input, the method 
  `LanguageDetector.computeLanguageConfidenceValues` computed wrong values
  under certain circumstances. (#87)
- When `Lingua` was used in projects with an explictly set Kotlin version
  which differed from Lingua's implicitly set version in the Gradle script,
  several errors occurred during runtime. By explicitly setting Lingua's
  Kotlin version, these errors are now hopefully gone. (#88, #89)
- Errors in the rule engine for the Latvian language have been resolved. (#92)

## Lingua 1.0.3 (released on 15 Oct 2020)

### Bug Fixes

- When two languages had exactly the same confidence values, 
one of them was erroneuosly removed from the result map.
Thanks to @mmedek for reporting this bug. (#72)
- There was still a problem with the classification of texts
consisting of certain alphabets.
Thanks to @nicolabertoldi for reporting this bug. (#76)
- The language detection for Spanish did not take the
rarely used accented characters á, é, í, ó, ú and ü
into account.
Thanks to @joeporter for reporting this bug. (#73)
- A bug in the rule engine led to weak detection accuracy
for Macedonian and Serbian. This has been fixed.

### Other changes

- The Kotlin compiler and runtime have been updated to version 1.4.
This includes the current stable release 1.0.0 of the 
kotlinx-serialization framework.
- The accuracy report files have been moved to their own Gradle source set.
This allows separate compilation of unit tests and accuracy report tests,
leading to more flexible and slightly faster compilation.

## Lingua 1.0.2 (released on 09 Aug 2020)

### Bug Fixes

- The language mapping for character *ë* was incorrect which has been fixed.
Thanks to @sandernugterenedia for reporting this bug. (#66)
- The implementation of `LanguageDetector` made use of functionality that was
introduced in Java 8 which made the library unusable for Java 6 and 7.
Thanks to @levant916 for reporting this bug. (#69)
- The [Gradle shadow plugin](https://imperceptiblethoughts.com/shadow/) has been
added so that `./gradlew jarWithDependencies` produces a jar file whose dependencies
do not conflict anymore with the same dependencies of different versions in the same project. (#67)

## Lingua 1.0.1 (released on 04 Jul 2020)

### Bug Fixes

- If no ngram probabilities were found for a given input text, a NullPointerException would be thrown.
Thanks to @fsonntag for finding and fixing this bug. (#63)

## Lingua 1.0.0 (released on 24 Jun 2020)

### Languages


src/python-scripts/draw_accuracy_plots.py
# Copyright © 2018-today Peter M. Stahl pemistahl@gmail.com
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either expressed or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import matplotlib
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns
from math import floor
from matplotlib.patches import Patch

matplotlib.use('TkAgg')
sns.set()
sns.set_style('whitegrid')


class AccuracyPlotDrawer(object):
    __dpi = 40
    __ticks_fontsize = 35
    __label_fontsize = 38
    __title_fontsize = 45
    __fontweight = 'bold'
    __hue = 'classifier'
    __grid_color = '#474747'
    __plot_filepath = 'images/plots/'
    __plot_titles = ('Single Word', 'Word Pair', 'Sentence', 'Average')
    __plot_title_suffix = 'Detection Performance'
    __column_prefixes = ('single-words', 'word-pairs', 'sentences', 'average')
    __column_suffixes = ('optimaize', 'opennlp', 'tika', 'lingua')
    __legend_labels = ('Optimaize 0.6', 'OpenNLP 1.9.4', 'Tika 2.1.0', 'Lingua 1.1.1')
    __hatches = ('/', '+', '.', 'O')
    __palette = ('#b259ff', '#ff6347', '#ffc400', '#41c46b')
    __ticks = np.arange(0, 101, 10)
    __legend_handles = [Patch(facecolor=color, edgecolor='black', label=label, hatch=hatch)
                        for color, label, hatch in zip(__palette, __legend_labels, __hatches)]

    def __init__(self, filepath):
        self.__dataframe = self.__read_into_dataframe(filepath)

    def __read_into_dataframe(self, filepath):
        frame = pd.read_csv(filepath_or_buffer=filepath)
        return pd.melt(frame=frame, id_vars='language', value_name='accuracy', var_name=self.__hue)

    def draw_all_barplots(self):
        for title, prefix in zip(self.__plot_titles, self.__column_prefixes):
            suffixed_title = title + ' ' + self.__plot_title_suffix
            columns = [prefix + '-' + suffix for suffix in self.__column_suffixes]
            self.draw_barplot(columns, suffixed_title, xlim=(0, 100), filename='barplot-' + prefix + '.png')

    def draw_all_boxplots(self):
        for title, prefix in zip(self.__plot_titles, self.__column_prefixes):
            suffixed_title = title + ' ' + self.__plot_title_suffix
            columns = [prefix + '-' + suffix for suffix in self.__column_suffixes]
            self.draw_boxplot(columns, suffixed_title, ylim=(0, 100), filename='boxplot-' + prefix + '.png')

    def draw_barplot(self, columns, title, xlim, filename):
        row_filter = self.__dataframe[self.__hue].isin(columns)
        data = self.__dataframe[row_filter]

        plt.figure(figsize=(16, 100))
        plt.title(title + '\n', fontsize=self.__title_fontsize, fontweight=self.__fontweight)
        plt.xticks(fontsize=self.__ticks_fontsize, ticks=self.__ticks)
        plt.yticks(fontsize=self.__ticks_fontsize)
        plt.grid(color=self.__grid_color)

        axes = sns.barplot(data=data, x='accuracy', y='language', hue=self.__hue, palette=self.__palette)

        axes.set_xlabel('Accuracy (%)\n', fontsize=self.__label_fontsize, fontweight=self.__fontweight)
        axes.set_ylabel('Language', fontsize=self.__label_fontsize, fontweight=self.__fontweight)
        axes.set_xlim(xlim)
        axes.xaxis.tick_top()
        axes.xaxis.set_label_position('top')
        axes.tick_params(axis='both', which='major', labelsize=self.__label_fontsize)
        axes.tick_params(axis='both', which='minor', labelsize=self.__label_fontsize)
        axes.legend(handles=self.__legend_handles, fontsize=28, loc='upper left')

        language_count = len(axes.patches) / len(self.__legend_labels)
        for i, current_bar in enumerate(axes.patches):
            current_bar.set_edgecolor(self.__grid_color)
            current_bar.set_hatch(self.__hatches[floor(i / language_count)])

        plt.tight_layout()
        plt.savefig(self.__plot_filepath + filename, dpi=self.__dpi)

    def draw_boxplot(self, columns, title, ylim, filename):
        row_filter = self.__dataframe[self.__hue].isin(columns)
        data = self.__dataframe[row_filter]

        plt.figure(figsize=(24, 12))
        plt.title(title, fontsize=self.__title_fontsize, fontweight=self.__fontweight)
        plt.xticks(fontsize=self.__ticks_fontsize)
        plt.yticks(fontsize=self.__ticks_fontsize, ticks=self.__ticks)
        plt.grid(self.__grid_color)

        axes = sns.boxplot(data=data, x='classifier', y='accuracy', linewidth=5, palette=self.__palette)

        axes.set_ylim(ylim)
        axes.set_xlabel('Classifier', fontsize=self.__label_fontsize, fontweight=self.__fontweight)
        axes.set_ylabel('Accuracy (%)', fontsize=self.__label_fontsize, fontweight=self.__fontweight)
        axes.set_xticklabels(self.__legend_labels)

        plt.tight_layout()
        plt.savefig(self.__plot_filepath + filename, dpi=self.__dpi)


drawer = AccuracyPlotDrawer(filepath='accuracy-reports/aggregated-accuracy-values.csv')
drawer.draw_all_barplots()
drawer.draw_all_boxplots()

print("All plots created successfully")
